### PR TITLE
Account Discovery

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/18104

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by yrliou** [comment]
Build error on unit tests currently btw.

**Comment by supermassive** [comment]
> Build error on unit tests currently btw.
There was gmock include missing. Fixed now.

File: components/brave_wallet/browser/keyring_service.cc [hunk 1]
```diff
@@ -1093,6 +1099,47 @@ void KeyringService::AddAccountForKeyring(const std::string& keyring_id,
...
+  if (!keyring)
+    return;
+  json_rpc_service_->GetTransactionCount(
+      keyring->GetDiscoveryAddress(discovery_account_index),
+      base::BindOnce(&KeyringService::OnGetTransactionCount,
+                     discovery_weak_factory_.GetWeakPtr(),
+                     discovery_account_index, attempts_left));
+}
+
+void KeyringService::OnGetTransactionCount(int discovery_account_index,
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/keyring_service.cc]: Fixed.

File: components/brave_wallet/browser/keyring_service.cc [hunk 2]
```diff
@@ -1093,6 +1099,47 @@ void KeyringService::AddAccountForKeyring(const std::string& keyring_id,
       keyring->GetAddress(accounts_num - 1), keyring_id);
 }
 
+void KeyringService::AddDiscoveryAccountsForKeyring(int discovery_account_index,
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/keyring_service.cc]: Fixed.

[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/keyring_service.cc]: Added comment to `KeyringService::RestoreWallet` method

**Review by yrliou — APPROVED** [review]
++ if CI is green, nice work.
Please update the test plan in the PR description for QA manual test steps and apply QA/Test-Plan-Specified label in the associated issue, thanks.

File: components/brave_wallet/browser/keyring_service.h [hunk 3]
```diff
@@ -37,11 +38,12 @@ class EthTransaction;
 class KeyringServiceUnitTest;
 class BraveWalletProviderImplUnitTest;
 class FilecoinKeyring;
+class JsonRpcService;
 
 // This class is not thread-safe and should have single owner
 class KeyringService : public KeyedService, public mojom::KeyringService {
  public:
-  explicit KeyringService(PrefService* prefs);
+  KeyringService(JsonRpcService* json_rpc_service, PrefService* prefs);
```
Comment (kylehickinson) [inline components/brave_wallet/browser/keyring_service.h:46]: Will require an update on iOS to pass in the rpc service like desktops factory: https://github.com/brave/brave-core/blob/5d6e4577ebe948efc17dec768265b59b2c3f443c/ios/browser/brave_wallet/keyring_service_factory.cc#L49

**Review by kylehickinson — CHANGES_REQUESTED** [review]

**Review by kylehickinson — APPROVED** [review]
iOS++

**Pull request merged**